libs/services/src/SlippageService/SlippageServiceMain.ts

```diff
@@ -15,35 +23,124 @@ export class SlippageServiceMain implements SlippageService {
+    if (volatilityOnExpiration === null) {
+      return MAX_SLIPPAGE_BPS;
+    }
+
+    // Return the slippage based on the volatility
+    return this.getSlippageBpsFromVolatility(volatilityOnExpiration);
+  }
+
+  private getSlippageBpsFromVolatility(volatility: number): Bps {
+    const slippageBps = Math.ceil(volatility * 10_000);
+
+    if (slippageBps < MIN_SLIPPAGE_BPS) {
+      return MIN_SLIPPAGE_BPS;
+    }
+
+    if (slippageBps > MAX_SLIPPAGE_BPS) {
+      return MAX_SLIPPAGE_BPS;
+    }
+
+    return slippageBps;
+  }
+
+  private async getMaxVolatilityOnExpiration({
+    order,
+    chainId,
+    baseTokenAddress,
+    quoteTokenAddress,
+  }: GetSlippageBpsParams) {
+    // Get the 5min standard deviation for the quote token (~288 points, 5min apart)
+    const [volatilityQuote, volatilityBase] = await Promise.all([
+      this.getVolatilityOnSettlement(chainId, quoteTokenAddress, order),
+      this.getVolatilityOnSettlement(chainId, baseTokenAddress, order),
     ]);
 
-    return Math.max(slippageQuoteToken, slippageBaseToken);
+    if (volatilityQuote === null || volatilityBase === null) {
+      return null;
+    }
+
+    return Math.max(volatilityQuote, volatilityBase);
   }
 
-  private async getSlippageForToken(tokenAddress: string): Promise<Bps> {
+  /**
+   * Get the volatility of the asset in some time (enough for a solver to execute a solvable order)
+   *
+   * @param chainId
+   * @param tokenAddress
+   *
+   * @returns volatility in decimal format
+   */
+  private async getVolatilityOnSettlement(
+    chainId: SupportedChainId,
+    tokenAddress: string,
+    order?: OrderForSlippageCalculation
+  ): Promise<number | null> {
     const prices = await this.usdRepository.getUsdPrices(
-      SupportedChainId.MAINNET,
+      chainId,
       tokenAddress,
-      'daily'
+      '5m'
     );
 
     if (!prices) {
-      return DEFAULT_SLIPPAGE_BPS;
+      return null;
     }
 
-    const todayPrice = prices[prices.length - 1].price;
-    const yesterdayPrice = prices[prices.length - 2].price;
-    const bps = Math.abs(
-      ((todayPrice - yesterdayPrice) / yesterdayPrice) * 10000
+    // Get price of the token
+    const usdPrice = await this.usdRepository.getUsdPrice(
+      chainId,
+      tokenAddress
+    );
+
+    if (!usdPrice) {
+      return null;
+    }
+
+    // Calculate the average of the prices (in USD)
+    const averagePrice =
+      prices.reduce((acc, price) => acc + price.price, 0) / prices.length;
+
+    // Calculate price differences between each price point and the average
+    const priceDifferences = prices.map((price) => price.price - averagePrice);
+
+    // Square the differences
+    const squaredDifferences = priceDifferences.map(
+      (difference) => difference ** 2
     );
 
-    return Math.ceil(bps);
+    // Calculate the variance
+    const variance =
+      squaredDifferences.reduce((acc, difference) => acc + difference, 0) /
+      prices.length;
```

Comment:
Nitpick and tiny optimization suggestion: `variance` can be calculated in a single loop, without the intermediate `priceDifferences` and `squaredDifferences` maps.

Something like:

```suggestion
    // Calculate the variance
    const variance = prices.reduce((acc, price) => {
      // Calculate price differences between the price point and the average
      const difference = price.price - averagePrice
      // Square the difference
      const squaredDifference = difference ** 2
      // Sum squared differences
      return acc + squaredDifference
    }, 0) / prices.length
```

Reply:
I tried to commit this, but said the code is outdated and wouldn't let me. I just manually added your changes